File: leetcode/python/3_longest_substring_without_repeating.py
from collections import defaultdict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charPos = dict[str, int]()
        left: int = 0
        right: int = 0
        length: int = 0

        while right < len(s):
            c = s[right]

            index = charPos.get(c)
            if index != None and index >= left:
                left = index + 1
            
            length = max(length, right - left + 1)
            charPos[c] = right
            right += 1
        
        return length

if __name__ == '__main__':
    s = Solution()
    print(s.lengthOfLongestSubstring("abcabcbb"))
    print(s.lengthOfLongestSubstring("bbbbb"))
    print(s.lengthOfLongestSubstring("pwwkew"))
    print(s.lengthOfLongestSubstring(""))

File: leetcode/c++/3_longest_substring_without_repeating.cpp
#include <map>
#include <algorithm>
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char,int> lastChar {};
        int res {0};
        int l {0};
        for (int i = 0; i < s.length(); ++i) {
            auto c {s[i]};
            if (lastChar.find(c) != lastChar.end()) {
                l = max(l, lastChar[c]);
            }
            res = max(res, i - l + 1);
            lastChar[c] = i;
        }
        return res;
    }
};